api/pipeline_spec.proto

```diff
@@ -0,0 +1,420 @@
+syntax = "proto3";
+
+package ml_pipelines;
+
+import "google/protobuf/any.proto";
+
+// The spec of a pipeline.
+message PipelineSpec {
+  // The metadata of the pipeline.
+  PipelineInfo pipeline_info = 1;
+
+  // A list of pipeline tasks, which form a DAG.
+  repeated PipelineTaskSpec tasks = 2;
+
+  // The deployment config of the pipeline.
+  // The deployment config can be extended to provide platform specific configs.
+  // The supported config is [PipelineDeploymentConifg]().
+  google.protobuf.Any deployment_config = 3;
+
+  // The version of the sdk, which compiles the spec.
+  string sdk_version = 4;
+
+  // The version of the schema.
+  string schema_version = 5;
+
+  // The definition of the runtime parameter.
+  message RuntimeParameter {
+    // Required field. The type of the runtime parameter.
+    PrimitiveType.PrimitiveTypeEnum type = 1;
+    // Optional field. Default value of the runtime parameter. If not set and
+    // the runtime parameter value is not provided during runtime, an error will
+    // be raised.
+    Value default_value = 2;
+  }
+  // The runtime parameters of the pipeline. Keyed by parameter name.
+  map<string, RuntimeParameter> runtime_parameters = 6;
+}
+
+// The spec of task inputs.
+message TaskInputsSpec {
+  // The specification of a task input artifact.
+  message InputArtifactSpec {
+    // The name of the upstream task which produces the output that matches with
+    // the `output_artifact_key`.
+    string producer_task = 1;
+
+    // The key of [TaskOutputsSpec.artifacts][] map of the producer task.
+    string output_artifact_key = 2;
+  }
+
+  // Represents an input parameter. The value can be taken from an upstream
+  // task's output parameter (if specifying `producer_task` and
+  // `output_parameter_key`, or it can be a runtime value, which can either be
+  // determined at compile-time, or from a pipeline parameter.
+  message InputParameterSpec {

```

Comment:
In KFP we usually use term "Argument" to clearly distinguish between input definitions and input arguments/values.

Reply:
We currently don't include the input definitions in the schema. Do you have a use case for that?

Reply:
>We currently don't include the input definitions in the schema.

I just find it a bit confusing that `TaskOutputsSpec` describes output definitions, but `TaskInputsSpec` doesn't describe input definitions.

Argo has nice separation with `Template`' having `.inputs` and `.outputs` while `DagTask` having `.arguments`.

Reply:
We don't currently need the input definition in the IR, which is more useful in component spec. Do you have any use case which will need the input definition in the pipeline IR?